source/Engine.cpp

```diff
@@ -870,9 +866,27 @@ void Engine::Step(bool isActive)
 	// Draw crosshairs on any minables in range of the flagship's scanners.
 	double scanRange = flagship ? 100. * sqrt(flagship->Attributes().Get("asteroid scan power")) : 0.;
 	if(flagship && scanRange && !flagship->IsHyperspacing())
+	{
+		// Decide before looping whether or not to catalog asteroids.  This
+		// results in cataloging in-range asteroids roughly 3 times a second.
+		bool shouldCatalogAsteroids = (!isAsteroidCatalogComplete && !Random::Int(20));
+		bool scanComplete = true;
 		for(const shared_ptr<Minable> &minable : asteroids.Minables())
 		{
 			Point offset = minable->Position() - center;
+
+			// Autocatalog asteroid: Record that the player knows this type of asteroid is available here.
+			if(shouldCatalogAsteroids && !asteroidsScanned.count(minable->Name()))
+			{
+				scanComplete = false;
+				if(!Random::Int(10) && (minable->Position() - flagship->Position()).Length() <= scanRange)
```

Comment:
Looking at line 876, that is already the position calculated. Change that to calculating the length of the offset to only do it once, and use it here on top of using it at line 890.
You're calculating the same thing twice.

If there is a flagship, center = flagship->Position() 100% (line 456)

Reply:
I opened a PR (https://github.com/samrocketman/endless-sky/pull/2) to your branch, merging it would resolve the small performance issues I still have